File: app/src/main/java/com/wusy/serialportproject/bean/EnvironmentalDetector.java
package com.wusy.serialportproject.bean;

import com.orhanobut.logger.Logger;

import java.util.ArrayList;

public class EnvironmentalDetector {
    private int PM2_5;//PM2.5
    private float temp;//温度
    private float humidity;//湿度
    private int CO2;//CO2
    private int PM2_5OutDoor;//室外PM2.5
    private double formaldehyde;//甲醛
    private double TVOC;//TVOC
    ArrayList<String> hexs;

    public EnvironmentalDetector(String data){
        data=data.substring(6);
        hexs=new ArrayList<>();
        for (int i=0;i<data.length()/4;i++){
            hexs.add(data.substring(i*4,i*4+4));
        }
        try{
            this.PM2_5=Integer.parseInt(hexs.get(0),16);
            this.temp=((float)Integer.parseInt(hexs.get(1),16))/10;
            this.humidity=((float)Integer.parseInt(hexs.get(2),16))/10;
            this.CO2=Integer.parseInt(hexs.get(3),16);
            this.TVOC=((double)Integer.parseInt(hexs.get(4),16))/100;
            this.PM2_5OutDoor=Integer.parseInt(hexs.get(6),16);
            this.formaldehyde=((double)Integer.parseInt(hexs.get(7),16))/1000;
        }catch (Exception e){
            Logger.e("解析环境数据发生错误",e);
        }

    }

    public int getPM2_5() {
        return PM2_5;
    }

    public void setPM2_5(int PM2_5) {
        this.PM2_5 = PM2_5;
    }

    public float getTemp() {
        return temp;
    }

    public void setTemp(float temp) {
        this.temp = temp;
    }

    public float getHumidity() {
        return humidity;
    }

    public void setHumidity(float humidity) {
        this.humidity = humidity;
    }

    public int getCO2() {
        return CO2;
    }

    public void setCO2(int CO2) {
        this.CO2 = CO2;
    }

    public int getPM2_5OutDoor() {
        return PM2_5OutDoor;
    }

    public void setPM2_5OutDoor(int PM2_5OutDoor) {
        this.PM2_5OutDoor = PM2_5OutDoor;
    }

    public double getFormaldehyde() {
        return formaldehyde;
    }

    public void setFormaldehyde(double formaldehyde) {
        this.formaldehyde = formaldehyde;
    }

    public double getTVOC() {
        return TVOC;
    }

    public void setTVOC(double TVOC) {
        this.TVOC = TVOC;
    }

    public ArrayList<String> getHexs() {
        return hexs;
    }

    public void setHexs(ArrayList<String> hexs) {
        this.hexs = hexs;
    }
}

File: app/src/main/java/com/wusy/serialportproject/bean/EnvironmentalDetector.kt
package com.wusy.serialportproject.bean

import com.orhanobut.logger.Logger
import com.wusy.serialportproject.devices.Ate24V
import com.wusy.serialportproject.devices.BaseDevices
import com.wusy.serialportproject.devices.EnvQ3

import java.util.ArrayList
import java.util.HashMap

class EnvironmentalDetector(data: String, devices: BaseDevices) {
    var pM2_5: Int = 0//PM2.5
    var temp: Float = 0f//温度
    var humidity: Float = 0.toFloat()//湿度
    var cO2: Int = 0//CO2
    var pM2_5OutDoor: Int = 0//室外PM2.5
    var formaldehyde: Double = 0.toDouble()//甲醛
    var tvoc: Double = 0.toDouble()//TVOC
    var hexs: ArrayList<String>? = null
    var pM2_5Level:Int=0//PM2.5等级
    var AQI:Int=0
    init {
        when (devices.name) {
            "EnvQ3" -> {
                val map=(devices as EnvQ3).parseData(data)
                this.pM2_5 = map["PM2.5"] as Int
                this.temp = map["temp"] as Float
                this.humidity = map["humidity"] as Float
                this.cO2 = map["CO2"] as Int
                this.tvoc = map["TVOC"] as Double
                this.pM2_5OutDoor =map["PM2.5OutDoor"] as Int
                this.formaldehyde = map["formaldehyde"] as Double
                this.hexs = map["hexs"] as ArrayList<String>
            }
            "Ate24V"->{
                val map=(devices as Ate24V).parseData(data)
                this.pM2_5 = (map["PM2.5"]?:0) as Int
                this.temp = (map["temp"]?:0f) as Float
                this.humidity = (map["humidity"]?:0f) as Float
                this.cO2 = (map["CO2"]?:0 )as Int
                this.tvoc = (map["TVOC"]?:0.0 )as Double
                this.pM2_5Level = (map["PM2.5污染等级"] ?:0)as Int
                this.hexs = (map["hexs"]?:ArrayList<String>()) as ArrayList<String>
                this.AQI=(map["AQI"]?:0) as Int
            }
        }


    }
}
